Plugins/iOS/EngageBySailthruWrapper.mm

```diff
@@ -0,0 +1,314 @@
+#import "EngageBySailthruWrapper.h"
+#import <Marigold/Marigold.h>
+
+const char *MAR_ST_ENGAGE_BY_ST = "EngageBySailthru";
+const char *MAR_ST_RECEIVE_ERROR = "ReceiveError";
+const char *MAR_ST_RECEIVE_PROFILE_VARS = "ReceiveProfileVars";
+const char *MAR_ST_RECEIVE_UNWRAPPED_LINK = "ReceiveUnwrappedLink";
+
+@interface MARPurchase ()
+
+- (nullable instancetype)initWithDictionary:(nonnull NSDictionary *)dictionary;
+
+@end
+
+static EngageBySailthruWrapper * _sharedInstance = nil;
+static dispatch_once_t onceSharedPredicate = 0;
+
+@interface EngageBySailthruWrapper ()
+/*
+ * We need to hold these blocks to make sure they are not released
+ * by ARC before they're executed and the scope variable are destroyed.
+ * Seems to be unique to the Unity runtime and Objective-C++.
+ */
+@property (nonatomic, copy) void (^errorBlock)(NSError *error);
+@property (nonatomic, copy) void (^profileVarsBlock)(NSDictionary<NSString *, id> * _Nullable, NSError * _Nullable);
+
++ (EngageBySailthruWrapper *)shared;
+
+@end
+
+@implementation EngageBySailthruWrapper
+
+# pragma mark - C Methods
+
+# pragma mark Page Tracking
+
+void _trackPageview (const char *url, const char *tags) {
+    NSString *tagsString = nil;
+    if (tags) {
+        tagsString = [NSString stringWithUTF8String:tags];
+    }
+    [[EngageBySailthruWrapper shared] trackPageview:[NSString stringWithUTF8String:url] withTags:tagsString];
+
+}
+
+void _trackImpression (const char *sectionId, const char *urls) {
+    NSString *urlsString = nil;
+    if (urls) {
+        urlsString = [NSString stringWithUTF8String:urls];
+    }
+    [[EngageBySailthruWrapper shared] trackImpression:[NSString stringWithUTF8String:sectionId] withUrls:urlsString];
```

Comment:
If `urlsString` is _not_ defined at line 49, due to 48 being a nullptr, does it still make sense to pass this onto the wrapper? Or should we return early before this point?

Reply:
I guess this also applies in other places where you've done the NULL check

Reply:
In trackImpression the urls are optional so it's OK to pass through a null just as long as we don't attempt to create an NSString with it, because that will throw an error.